Index: fetch-spl.sh
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- fetch-spl.sh	(revision 745a97685408ff4249d96333a7885cd214639b70)
+++ fetch-spl.sh	(revision 97144cdb8e9cb4d83943b0b5898d08f57844a4dd)
@@ -40,4 +40,8 @@
 fetch_program token 1.0.0 TokenSVp5gheXUvJ6jGWGeCsgPKgnE3YgdGKRVCMY9o BPFLoader1111111111111111111111111111111111
 fetch_program memo  1.0.0 Memo1UhkJRfHyvLMcVucJwxXeuD728EqVDDwQDxFMNo BPFLoader1111111111111111111111111111111111

+# Add bridge
+genesis_args+=(--bpf-program Bridge1p5gheXUvJ6jGWGeCsgPKgnE3YgdGKRVCMY9o BPFLoader2111111111111111111111111111111111 /opt/solana/deps/spl_bridge.so)
+genesis_args+=(--bpf-program EevaaBridgeeXUvJ6jGWGeCsgPKgnE3YgdGKRVCMY9o BPFLoader2111111111111111111111111111111111 /opt/solana/deps/ee_vaa_program.so)
+
 echo "${genesis_args[@]}" > spl-genesis-args.sh

 echo
