diff --git a/ethereum/contracts/bridge/Bridge.sol b/ethereum/contracts/bridge/Bridge.sol
index 3f06757f..6c70e6fa 100644
--- a/ethereum/contracts/bridge/Bridge.sol
+++ b/ethereum/contracts/bridge/Bridge.sol
@@ -58,7 +58,7 @@ contract Bridge is BridgeGovernance, ReentrancyGuard {
 
         sequence = wormhole().publishMessage{
             value : msg.value
-        }(nonce, encoded, 15);
+        }(nonce, encoded, 1);
     }
 
     function wrapAndTransferETH(uint16 recipientChain, bytes32 recipient, uint256 arbiterFee, uint32 nonce) public payable returns (uint64 sequence) {
@@ -173,7 +173,7 @@ contract Bridge is BridgeGovernance, ReentrancyGuard {
 
         sequence = wormhole().publishMessage{
             value : callValue
-        }(nonce, encoded, 15);
+        }(nonce, encoded, 1);
     }
 
     function updateWrapped(bytes memory encodedVm) external returns (address token) {
diff --git a/ethereum/contracts/nft/NFTBridge.sol b/ethereum/contracts/nft/NFTBridge.sol
index abf25a28..78f9a781 100644
--- a/ethereum/contracts/nft/NFTBridge.sol
+++ b/ethereum/contracts/nft/NFTBridge.sol
@@ -95,7 +95,7 @@ contract NFTBridge is NFTBridgeGovernance {
 
         sequence = wormhole().publishMessage{
             value : callValue
-        }(nonce, encoded, 15);
+        }(nonce, encoded, 1);
     }
 
     function completeTransfer(bytes memory encodedVm) public {
