--- node_modules/truffle-plugin-verify/contants-old.js	2022-01-28 12:44:53.140709519 +0000
+++ node_modules/truffle-plugin-verify/constants.js	2022-01-28 12:45:16.800710171 +0000
@@ -3,13 +3,25 @@
   3: 'https://api-ropsten.etherscan.io/api',
   4: 'https://api-rinkeby.etherscan.io/api',
   5: 'https://api-goerli.etherscan.io/api',
+  10: 'https://api-optimistic.etherscan.io/api',
   42: 'https://api-kovan.etherscan.io/api',
   56: 'https://api.bscscan.com/api',
+  69: 'https://api-kovan-optimistic.etherscan.io/api',
   97: 'https://api-testnet.bscscan.com/api',
   128: 'https://api.hecoinfo.com/api',
   137: 'https://api.polygonscan.com/api',
   250: 'https://api.ftmscan.com/api',
   256: 'https://api-testnet.hecoinfo.com/api',
+  1284: 'https://api-moonbeam.moonscan.io/api',
+  1285: 'https://api-moonriver.moonscan.io/api',
+  1287: 'https://api-moonbase.moonscan.io/api',
+  4002: 'https://api-testnet.ftmscan.com/api',
+  42161: 'https://api.arbiscan.io/api',
+  43113: 'https://api-testnet.snowtrace.io/api',
+  43114: 'https://api.snowtrace.io/api',
+  421611: 'https://api-testnet.arbiscan.io/api',
+  1313161554: 'https://api.aurorascan.dev/api',
+  1313161555: 'https://api-testnet.aurorascan.dev/api',
   80001: 'https://api-testnet.polygonscan.com/api'
 }
 
@@ -18,16 +28,27 @@
   3: 'https://ropsten.etherscan.io/address',
   4: 'https://rinkeby.etherscan.io/address',
   5: 'https://goerli.etherscan.io/address',
+  10: 'https://optimistic.etherscan.io/address',
   42: 'https://kovan.etherscan.io/address',
   56: 'https://bscscan.com/address',
+  69: 'https://kovan-optimistic.etherscan.io/address',
   97: 'https://testnet.bscscan.com/address',
   128: 'https://hecoinfo.com/address',
   137: 'https://polygonscan.com/address',
   250: 'https://ftmscan.com/address',
   256: 'https://testnet.hecoinfo.com/address',
+  1284: 'https://moonbeam.moonscan.io/address',
+  1285: 'https://moonriver.moonscan.io/address',
+  1287: 'https://moonbase.moonscan.io/address',
+  4002: 'https://testnet.ftmscan.com/address',
+  42161: 'https://arbiscan.io/address',
+  43113: 'https://testnet.snowtrace.io/address',
+  43114: 'https://snowtrace.io/address',
+  421611: 'https://testnet.arbiscan.io/address',
+  1313161554: 'https://aurorascan.dev/address',
+  1313161555: 'https://testnet.aurorascan.dev/address',
   80001: 'https://mumbai.polygonscan.com/address'
 }
-
 const RequestStatus = {
   OK: '1',
   NOTOK: '0'
